### ISSUE TITLE
Executing a shell to run printf for extracting env vars no longer works with 2.5.0

### ISSUE THREAD

**Issue by lvh** [fish-shell/fish-shell#3809]
- [*] Have you checked if problem occurs with [fish 2.5.0](/fish-shell/fish-shell/releases/tag/2.5.0)? 
- [*] Tried fish without third-party customizations *(check `sh -c 'env HOME=$(mktemp -d) fish'`)*?

**fish version installed** *(`fish --version`)*: `fish, version 2.5.0`

**OS/terminal used**: macOS Sierra 10.12.4 Beta (16E144f) + iTerm2

## Reproduction steps

Not entirely sure if this is a spacemacs bug, or a thefuck bug, or a fish bug, but, since upgrading from 2.4.0 to 2.5.0, starting emacs (spacemacs) gives me:

```
Error (use-package): exec-path-from-shell :init: Expected printf output from shell, but got: ""
Warning (initialization): An error occurred while loading ‘/Users/lvh/.emacs.d/init.el’:

error: Expected printf output from shell, but got: ""

To ensure normal operation, you should investigate and remove the
cause of the error in your initialization file.  Start Emacs with
the ‘--debug-init’ option to view a complete error backtrace.
```

This reliably goes away when I downgrade to 2.4.0.

My `fish.config` is:

```fish
function interactive_setup
  . (pyenv init -|psub);
  . (pyenv virtualenv-init -|psub);
  . (rbenv init -|psub);
  eval (thefuck --alias | tr '\n' ';')
  . /usr/local/share/autojump/autojump.fish
end

status --is-interactive; and interactive_setup
```

In an attempt to find the problem, I have narrowed it down by enabling/disabling all of the lines in `interactive_setup`, and found out that the cuplrit was `eval (thefuck --alias | tr '\n' ';')`, the somewhat unfortunately named `thefuck`, which does command correction. Just so you have an idea what it produces:

```
lvh@plotka ~/.c/f/functions> thefuck --alias
function fuck -d "Correct your previous console command"
  set -l fucked_up_command $history[1]
  env TF_ALIAS=fuck PYTHONIOENCODING=utf-8 thefuck $fucked_up_command | read -l unfucked_command
  if [ "$unfucked_command" != "" ]
    eval $unfucked_command
    builtin history delete --exact --case-sensitive -- $fucked_up_command
    builtin history merge ^ /dev/null
  end
end
```

I traced the emacs code down as well down to https://github.com/purcell/exec-path-from-shell. It's very strange that the returned output should be empty; it should have two NUL bytes in it, even if the $PATH and $MANPATH variables are empty. Confusingly, when I do that in an iTerm2 session it works fine:

```
lvh@plotka ~/.c/f/functions> printf "$PATH\000$MANPATH"
/usr/local/Cellar/pyenv-virtualenv/1.0.0/shims /Users/lvh/.pyenv/shims /usr/local/bin /usr/bin /bin /usr/sbin /sbin /opt/X11/bin /usr/local/MacGPG2/bin /Library/TeX/texbin /Applications/Wireshark.app/Contents/MacOS⏎
```

Several people have reported this elsewhere: https://github.com/purcell/exec-path-from-shell/issues/56

**Comment by lvh** [comment]
Oops. Clicked submit too soon. Sorry for the originally awful title.

**Comment by faho** [comment]
It's a bit hard to follow what actually is executed here. I installed exec-path-from-shell, called `exec-path-from-shell-initialize` and then `exec-path-from-shell-copy-env` with "PATH" as the argument, and it worked.

When I instead use "$PATH" (note the $), it fails with the following message:

>Non-zero exit code from shell /usr/bin/fish invoked with args ("-l" "-i" "-c" "sh -c /usr/bin/printf\\ \\’__RESULT\\\\000\\%s\\’\\ \\\"\\$\
Uses keymap ‘\\$PATH-53b1fb55843b8ce75befa12cb767e712\\’, which is not currently defined.
\\\"").  Output was:
"sh: ${$PATH-53b1fb55843b8ce75befa12cb767e712}: Falsche Variablenersetzung.
"

As you can see, fish is called as an interactive login shell, so your interactive setup _is_ called. What happens when you remove that?

(I'll also have to check old issues, since I'm quite sure we've had one with this package in the past)

**Comment by faho** [comment]
Also important is that it's not _our_ printf being called, since exec-path-from-shell doesn't handle fish's output format. Instead, it asks fish to launch `sh -c` to then printf $PATH.

(All of this is done in order to get the user's configured $PATH, not the system default for graphical applications - this is unnecessary when using emacs inside a terminal, but interesting otherwise)

**Comment by faho** [comment]
So, I've installed spacemacs.

It appears that (for some reason) it also uses exec-path-from-shell even inside of a terminal, though my configuration does not trigger the issue.

What does is _any_ call to `eval` in config.fish.

Which for one means, @lvh, there's an easy workaround for you, and it's cleaner and faster than what you have right now - use `thefuck --alias | source` instead of `eval`.

Most likely the problem here is the job-control stuff eval does that I've never quite understood.

**Comment by lvh** [comment]
@faho Whoa! That's a lot of effort. Thanks!

Sorry if I said something misleading re: interactive_setup being called. My understanding is that running it interactively is intentional; I think it does things like e.g. find your `ipython` binary; so if you use tricks like pyenv and you want the right venv to be activated automatically, it's important that it can ask a real shell.

**Comment by lvh** [comment]
`source` fixed the issue; although I'll be darned if I have a clue why. I'll update the documentation for `thefuck`.

**Comment by lvh** [comment]
I have fixed the instructions for `thefuck` so that hopefully it will affect fewer people in the future.

**Comment by faho** [comment]
It seems I cannot reproduce this with current master, only 2.5.0, so it seems to have been fixed. I'll close this issue unless you or anyone else can reproduce it.

This would be something for a 2.5.1 if we did it.

**Comment by lvh** [comment]
Sure! As long as there's a ticket with a workaround that has a lot of references to other tickets it should be easy for people to fix the issue :) Thanks again for your work on debugging this! To be clear: there's no reason to use `eval` with the weird `tr` step to begin with right?

**Comment by pbugnion** [comment]
For reference, I hit the same spacemacs / fish issue. In my case, the offending line was:

```
eval (python -m virtualfish)
```

Changing this to:

```
python -m virtualfish | source
```

seems to fix it.

**Comment by faho** [comment]
>there's no reason to use eval with the weird tr step to begin with right?

Yes, there's no reason here. The thing `eval` does is allow stuff like redirections and access to stdin and so. For these kinds of commands, you don't need that.

What the `tr` there does is allow that alias to span multiple commands - the \n will be swallowed by the command substitutions, which instead splits on those. Only `eval` doesn't care about how it's arguments are split - `eval $var` and `eval "$var"` are equivalent. (This is a mistake, but not something we want to change for fear of breaking too much stuff)

So replacing newlines with ";" turns a multi-line multi-command into a single-line multi-command expression, which eval allows.

`source` simply reads commands line-by-line to begin with.

**Comment by balbeko** [comment]
I've submitted PRs with fiexes to RVM and oh-my-fish, hope this would be merged there soon
